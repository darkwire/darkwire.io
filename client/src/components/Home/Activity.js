import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Message from 'components/Message'
import Username from 'components/Username'
import Notice from 'components/Notice'
import Zoom from 'utils/ImageZoom'
import { getObjectUrl } from 'utils/file'

import T from 'components/T'

class Activity extends Component {
  constructor(props) {
    super(props)

    this.state = {
      zoomableImages: [],
    }
  }

  getFileDisplay(activity) {
    const type = activity.fileType
    if (type.match('image.*')) {
      return (
        <img
          ref={c => this._zoomableImage = c}
          className="image-transfer zoomable"
          src={`data:${activity.fileType};base64,${activity.encodedFile}`}
          alt={`${activity.fileName} from ${activity.username}`}
          onLoad={this.handleImageDisplay.bind(this)}
        />
      )
    }
    return null
  }

  handleImageDisplay() {
    Zoom(this._zoomableImage)
    this.props.scrollToBottom()
  }

  getActivityComponent(activity) {
    switch (activity.type) {
      case 'TEXT_MESSAGE':
        return (
          <Message
            sender={activity.username}
            message={activity.text}
            timestamp={activity.timestamp}
          />
        )
      case 'USER_ENTER':
        return (
          <Notice>
            <div>
              <T data={{
                username: <Username key={0} username={activity.username} />
              }} path='userJoined'/>
            </div>
          </Notice>
        )
      case 'USER_EXIT':
        return (
          <Notice>
            <div>
              <T data={{
                username: <Username key={0} username={activity.username} />
              }} path='userLeft'/>
            </div>
          </Notice>
        )
      case 'TOGGLE_LOCK_ROOM':
        if (activity.locked) {
          return (
            <Notice>
              <div><T data={{
                username: <Username key={0} username={activity.username} />
              }} path='lockedRoom'/></div>
            </Notice>
          )
        } else {
          return (
            <Notice>
              <div><T data={{
                username: <Username key={0} username={activity.username} />
              }} path='unlockedRoom'/></div>
            </Notice>
          )
        }
      case 'NOTICE':
        return (
          <Notice>
            <div>{activity.message}</div>
          </Notice>
        )
      case 'CHANGE_USERNAME':
        return (
          <Notice>
            <div><T data={{
              oldUsername: <Username key={0} username={activity.currentUsername} />,
              newUsername: <Username key={1} username={activity.newUsername} />
            }} path='nameChange'/>
            </div>
          </Notice>
        )
      case 'USER_ACTION':
        return (
          <Notice>
            <div>&#42; <Username username={activity.username} /> {activity.action}</div>
          </Notice>
        )
      case 'RECEIVE_FILE':
        const downloadUrl = getObjectUrl(activity.encodedFile, activity.fileType)
        return (
          <div>
            <T data={{
              username: <Username key={0} username={activity.username} />,
            }} path='userSentFile'/>&nbsp;

            <a target="_blank" href={downloadUrl} rel="noopener noreferrer" download={activity.fileName}>
              <T data={{
                filename: activity.fileName,
              }} path='downloadFile'/>
            </a>
            {this.getFileDisplay(activity)}
          </div>
        )
      case 'SEND_FILE':
        const url = getObjectUrl(activity.encodedFile, activity.fileType)
        return (
          <Notice>
            <div>
              <T data={{
                filename: <a key={0} target="_blank" href={url} rel="noopener noreferrer" download={activity.fileName}>{activity.fileName}</a>,
              }} path='sentFile'/>&nbsp;
            </div>
            {this.getFileDisplay(activity)}
          </Notice>
        )
      default:
        return false
    }
  }

  render() {
    return (
      this.getActivityComponent(this.props.activity)
    )
  }
}

Activity.propTypes = {
  activity: PropTypes.object.isRequired,
  scrollToBottom: PropTypes.func.isRequired,
}

export default Activity;
