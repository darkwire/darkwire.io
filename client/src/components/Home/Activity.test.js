import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Activity from './Activity';
import { Provider } from 'react-redux';
import configureStore from 'store';

const store = configureStore();

//jest.mock('components/T'); // Need store

describe('Activity component', () => {
  it('should display', () => {
    const activity = {
      type: '',
    };
    const { asFragment } = render(
      <Provider store={store}>
        <Activity activity={activity} scrollToBottom={jest.fn()} />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should display TEXT_MESSAGE', () => {
    const activity = {
      type: 'TEXT_MESSAGE',
      username: 'alice',
      timestamp: new Date('2020-03-14T11:01:58.135Z').valueOf(),
      text: 'Hi!',
    };
    const { asFragment } = render(
      <Provider store={store}>
        <Activity activity={activity} scrollToBottom={jest.fn()} />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should display USER_ENTER', () => {
    const activity = {
      type: 'USER_ENTER',
      username: 'alice',
    };
    const { asFragment } = render(
      <Provider store={store}>
        <Activity activity={activity} scrollToBottom={jest.fn()} />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should display USER_EXIT', () => {
    const activity = {
      type: 'USER_EXIT',
      username: 'alice',
    };
    const { asFragment } = render(
      <Provider store={store}>
        <Activity activity={activity} scrollToBottom={jest.fn()} />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should display TOGGLE_LOCK_ROOM', () => {
    const activity = {
      type: 'TOGGLE_LOCK_ROOM',
      locked: true,
      username: 'alice',
    };
    const { asFragment, rerender } = render(
      <Provider store={store}>
        <Activity activity={activity} scrollToBottom={jest.fn()} />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();

    activity.locked = false;

    rerender(
      <Provider store={store}>
        <Activity activity={activity} scrollToBottom={jest.fn()} />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should display NOTICE', () => {
    const activity = {
      type: 'NOTICE',
      message: 'Hello world!',
    };
    const { asFragment } = render(
      <Provider store={store}>
        <Activity activity={activity} scrollToBottom={jest.fn()} />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should display CHANGE_USERNAME', () => {
    const activity = {
      type: 'CHANGE_USERNAME',
      currentUsername: 'alice',
      newUsername: 'alicette',
    };
    const { asFragment } = render(
      <Provider store={store}>
        <Activity activity={activity} scrollToBottom={jest.fn()} />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should display USER_ACTION', () => {
    const activity = {
      type: 'USER_ACTION',
      username: 'alice',
      action: 'did right!',
    };
    const { asFragment } = render(
      <Provider store={store}>
        <Activity activity={activity} scrollToBottom={jest.fn()} />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should display RECEIVE_FILE', () => {
    const activity = {
      type: 'RECEIVE_FILE',
      username: 'alice',
      fileName: 'alice.pdf',
      encodedFile: 'dGV4dGZpbGU=',
      fileType: 'text/plain',
    };
    global.URL.createObjectURL = jest.fn(data => `url:${data}`);

    const { asFragment } = render(
      <Provider store={store}>
        <Activity activity={activity} scrollToBottom={jest.fn()} />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should display RECEIVE_FILE with image', () => {
    global.URL.createObjectURL = jest.fn(data => `url:${data}`);

    const mockScrollToBottom = jest.fn();

    const activity = {
      type: 'RECEIVE_FILE',
      username: 'alice',
      fileName: 'alice.jpg',
      encodedFile: 'dGV4dGZpbGU=',
      fileType: 'image/jpg',
    };

    const { asFragment, getByAltText } = render(
      <Provider store={store}>
        <Activity activity={activity} scrollToBottom={mockScrollToBottom} />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();

    const image = getByAltText('alice.jpg from alice');
    fireEvent.load(image);
    expect(mockScrollToBottom).toHaveBeenCalled();
  });

  it('should display SEND_FILE', () => {
    global.URL.createObjectURL = jest.fn(data => `url:${data}`);
    const activity = {
      type: 'SEND_FILE',
      username: 'alice',
      fileName: 'alice.pdf',
      encodedFile: 'dGV4dGZpbGU=',
      fileType: 'text/plain',
    };

    const { asFragment } = render(
      <Provider store={store}>
        <Activity activity={activity} scrollToBottom={jest.fn()} />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
