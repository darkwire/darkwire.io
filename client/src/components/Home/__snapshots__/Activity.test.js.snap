// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Activity component should display 1`] = `<DocumentFragment />`;

exports[`Activity component should display CHANGE_USERNAME 1`] = `
<DocumentFragment>
  <div>
    <div
      class="info"
    >
      <div>
        <span>
          <span
            class="username"
            style="color: rgb(204, 241, 255);"
          >
            alice
          </span>
           changed their name to 
          <span
            class="username"
            style="color: rgb(235, 162, 242);"
          >
            alicette
          </span>
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Activity component should display NOTICE 1`] = `
<DocumentFragment>
  <div>
    <div
      class="info"
    >
      <div>
        Hello world!
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Activity component should display RECEIVE_FILE 1`] = `
<DocumentFragment>
  <div>
    <span>
      <span
        class="username"
        style="color: rgb(204, 241, 255);"
      >
        alice
      </span>
       sent you a file.
    </span>
     
    <a
      download="alice.pdf"
      href="url:[object Blob]"
      rel="noopener noreferrer"
      target="_blank"
    >
      <span>
        Download alice.pdf
      </span>
    </a>
  </div>
</DocumentFragment>
`;

exports[`Activity component should display RECEIVE_FILE with image 1`] = `
<DocumentFragment>
  <div>
    <span>
      <span
        class="username"
        style="color: rgb(204, 241, 255);"
      >
        alice
      </span>
       sent you a file.
    </span>
     
    <a
      download="alice.jpg"
      href="url:[object Blob]"
      rel="noopener noreferrer"
      target="_blank"
    >
      <span>
        Download alice.jpg
      </span>
    </a>
    <img
      alt="alice.jpg from alice"
      class="image-transfer zoomable"
      src="data:image/jpg;base64,dGV4dGZpbGU="
    />
  </div>
</DocumentFragment>
`;

exports[`Activity component should display SEND_FILE 1`] = `
<DocumentFragment>
  <div>
    <div
      class="info"
    >
      <div>
        <span>
          You sent 
          <a
            download="alice.pdf"
            href="url:[object Blob]"
            rel="noopener noreferrer"
            target="_blank"
          >
            alice.pdf
          </a>
        </span>
         
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Activity component should display TEXT_MESSAGE 1`] = `
<DocumentFragment>
  <div>
    <div
      class="chat-meta"
    >
      <span
        class="username"
        style="color: rgb(204, 241, 255);"
      >
        alice
      </span>
      <span
        class="muted timestamp"
      >
        11:01 AM
      </span>
    </div>
    <div
      class="chat"
    >
      <span
        class="Linkify"
      >
        Hi!
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Activity component should display TOGGLE_LOCK_ROOM 1`] = `
<DocumentFragment>
  <div>
    <div
      class="info"
    >
      <div>
        <span>
          <span
            class="username"
            style="color: rgb(204, 241, 255);"
          >
            alice
          </span>
           locked the room
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Activity component should display TOGGLE_LOCK_ROOM 2`] = `
<DocumentFragment>
  <div>
    <div
      class="info"
    >
      <div>
        <span>
          <span
            class="username"
            style="color: rgb(204, 241, 255);"
          >
            alice
          </span>
           unlocked the room
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Activity component should display USER_ACTION 1`] = `
<DocumentFragment>
  <div>
    <div
      class="info"
    >
      <div>
        * 
        <span
          class="username"
          style="color: rgb(204, 241, 255);"
        >
          alice
        </span>
         did right!
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Activity component should display USER_ENTER 1`] = `
<DocumentFragment>
  <div>
    <div
      class="info"
    >
      <div>
        <span>
          <span
            class="username"
            style="color: rgb(204, 241, 255);"
          >
            alice
          </span>
           joined
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Activity component should display USER_EXIT 1`] = `
<DocumentFragment>
  <div>
    <div
      class="info"
    >
      <div>
        <span>
          <span
            class="username"
            style="color: rgb(204, 241, 255);"
          >
            alice
          </span>
           left
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
